### ISSUE TITLE
HPC mirroring... sr3_cpost from within a tree being mirrored.

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarrac#141]
have a user report of strange failures in a supercomputer jobs that is failing to run sr3_cpost when inside a job that has the shim library active.  start by trying to reproduce.

```

git clone https://github.com/MetPX/sr_insects insects
cd insects/static*
./flow_setup.sh declare


```
results in:

```
blacklab% sr3 status 'post/*'
status: 
Component/Config                         Processes   Connection        Lag                Rates                                        
                                         State   Run Retry  msg data   LagMax  LagAvg  %rej     pubsub   messages     RxData     TxData 
                                         -----   --- -----  --- ----   ------  ------  ----   --------       ----     ------     ------ 
post/shim_f63                            stop    0/0          -          -         -     -          -        -
post/t_dd1_f00                           stop    0/0          -          -         -     -          -        -
post/t_dd2_f00                           stop    0/0          -          -         -     -          -        -
post/test2_f61                           stop    0/0          -          -         -     -          -        -
      Total Running Configs:   0 ( Processes: 0 missing: 0 stray: 0 )
                     Memory: uss:0 Bytes rss:0 Bytes vms:0 Bytes 
                   CPU Time: User:0.00s System:0.00s 
	   Pub/Sub Received: 0 msgs/s (0 Bytes/s), Sent:  0 msgs/s (0 Bytes/s)
	      Data Received: 0 Files/s (0 Bytes/s), Sent: 0 Files/s (0 Bytes/s) 
blacklab% 

```

cd into metpx-sr3c source dir:
```

export LD_PRELOAD="`pwd`/libsr3shim.so.1.0.0"
export SR_POST_CONFIG=~/.config/sr3/post/shim_f63.conf
bash # start a subshell with the shim active.
cd ~/sarra_de*
mkdir hoho

```
now an actual post:

```
blacklab% /bin/echo hoho >lala* 
2024-01-18 17:56:53,470 [INFO] logEvents option not implemented, ignored.
2024-01-18 17:56:53,470 [INFO] tls_rigour option not implemented, ignored.
2024-01-18 17:56:53,470 [INFO] amqp_consumer option not implemented, ignored.
2024-01-18 17:56:53,479 [INFO] published: { "pubTime":"20240118225653.47458286", "baseUrl":"ftp://anonymous@localhost:2121", "relPath":"hoho/lala_wrapper.txt", "topic":"v03.post.hoho", "identity":{  "method" : "sha512", "value" : "V/BeoaZp+hh0CbopUiR/1kckdUspp4oAnsDmhZhaIXVKb3gGPB5lrT62enDtl5lxQ/e+vO9GsuoRH1\nY+rlGhxQ=="  } , "mtime":"20240118225653.46532026", "atime":"20240118224459.97378987", "mode":"0664", "size":"5"}
blacklab% 

```
that was produced by forking an echo... I noticed built-in echo had no effect... harrumph.
next try a cpost

```

blacklab% sr3_cpost -c ~/.config/sr3/post/t_dd1_f00.conf -p lala*
2024-01-18 17:58:27,039 [INFO] logEvents option not implemented, ignored.
2024-01-18 17:58:27,039 [INFO] tls_rigour option not implemented, ignored.
2024-01-18 17:58:27,039 [INFO] amqp_consumer option not implemented, ignored.
2024-01-18 17:58:27,040 [ERROR] Environment variable not set: SAMPLEDATA
2024-01-18 17:58:27,040 [INFO] set option not implemented, ignored.
2024-01-18 17:58:27,040 [INFO] set option not implemented, ignored.
2024-01-18 17:58:27,040 [DEBUG] sr_cpost 3.23.11p3 settings: action=foreground hostname=blacklab config_name=t_dd1_f00 log_level=6
2024-01-18 17:58:27,040 [DEBUG] 	follow_symlinks=no realpath: Adjust=0 Filter=yes Post=no
2024-01-18 17:58:27,040 [DEBUG] 	force_polling=off sleep=0 expire=86400 housekeeping=300 sanity_log_dead=450 nodupe_ttl=0 nodupe_fileAgeMin=0 nodupe_fileAgeMax=2.592e+06
2024-01-18 17:58:27,040 [DEBUG] 	cache_file=off cache_basis=path, accept_unmatch=on messageRateMax=0
2024-01-18 17:58:27,040 [DEBUG] 	fileEvents=006f directory=(null) queuename=(null) logReject=on sum=s
2024-01-18 17:58:27,040 [DEBUG] 	statehost=no v2compatRenameDoublePost=no
2024-01-18 17:58:27,040 [DEBUG] 	message_ttl=0 post_exchange=xsarra post_exchange_split=0 post_exchangeSuffix=(null)
2024-01-18 17:58:27,040 [DEBUG] 	source=A to=localhost post_baseUrl=http://localhost:8090 post_baseDir=(null)
2024-01-18 17:58:27,040 [DEBUG] 	topicPrefix=v03 post_topicPrefix=v03.post, pid=-1
2024-01-18 17:58:27,040 [DEBUG] man sr3_cpost(1) for more information
2024-01-18 17:58:27,043 [DEBUG] 3.23.11p3 connected to post broker amqp://tfeed@localhost:5672/#xsarra
2024-01-18 17:58:27,043 [DEBUG] declaring exchange amqp://tfeed@localhost:5672/#xsarra
2024-01-18 17:58:27,044 [DEBUG] sr_config_read about to open: /home/peter/.config/sr3/default.conf
2024-01-18 17:58:27,044 [INFO] logEvents option not implemented, ignored.
2024-01-18 17:58:27,044 [INFO] tls_rigour option not implemented, ignored.
2024-01-18 17:58:27,044 [INFO] amqp_consumer option not implemented, ignored.
2024-01-18 17:58:27,044 [DEBUG] sr_config_read about to open: /home/peter/.config/sr3/post/shim_f63.conf
2024-01-18 17:58:27,044 [DEBUG] sr_shim 3.23.11p3 settings: action=foreground hostname=blacklab config_name=NONE log_level=5
2024-01-18 17:58:27,044 [DEBUG] 	follow_symlinks=no realpath: Adjust=0 Filter=yes Post=no
2024-01-18 17:58:27,044 [DEBUG] 	force_polling=off sleep=0 expire=86400 housekeeping=300 sanity_log_dead=450 nodupe_ttl=0 nodupe_fileAgeMin=0 nodupe_fileAgeMax=2.592e+06
2024-01-18 17:58:27,044 [DEBUG] 	cache_file=off cache_basis=path, accept_unmatch=on messageRateMax=0
2024-01-18 17:58:27,044 [DEBUG] 	fileEvents=006f directory=(null) queuename=(null) logReject=off sum=s
2024-01-18 17:58:27,044 [DEBUG] 	statehost=no v2compatRenameDoublePost=no
2024-01-18 17:58:27,044 [DEBUG] 	message_ttl=0 post_exchange=xs_tsource_shim post_exchange_split=0 post_exchangeSuffix=(null)
2024-01-18 17:58:27,044 [DEBUG] 	source=tsource to=localhost post_baseUrl=ftp://anonymous@localhost:2121 post_baseDir=/home/peter/sarra_devdocroot
2024-01-18 17:58:27,044 [DEBUG] 	topicPrefix=v03 post_topicPrefix=v03.post, pid=-1
2024-01-18 17:58:27,044 [DEBUG] man sr3_cpost(1) for more information
2024-01-18 17:58:27,044 [DEBUG] realpath_adjust 0, /home/peter/.cache/sr3/cpost/t_dd1_f00/i01.pid -> /home/peter/.cache/sr3/cpost/t_dd1_f00/i01.pid 
2024-01-18 17:58:27,049 [DEBUG] 3.23.11p3 connected to post broker amqp://tsource@localhost:5672/#xs_tsource_shim
2024-01-18 17:58:27,049 [DEBUG] declaring exchange amqp://tsource@localhost:5672/#xs_tsource_shim
2024-01-18 17:58:27,049 [ERROR] posting outside of post_baseDir (/home/peter/sarra_devdocroot) invalid path: /home/peter/.cache/sr3/cpost/t_dd1_f00/i01.pid
2024-01-18 17:58:27,049 [INFO] cpost 3.23.11p3 config: t_dd1_f00, pid: 620464, starting
2024-01-18 17:58:27,049 [DEBUG] sr_event_mask: 006f translated to inotify_event_mask: 20003cc
2024-01-18 17:58:27,049 [DEBUG] starting polling loop pass: 0
2024-01-18 17:58:27,051 [DEBUG] v03 body={ "pubTime" : "20240118T225827.04942486", "baseUrl" : "http://localhost:8090", "relPath" : "home/peter/sarra_devdocroot/hoho/lala_wrapper.txt", "identity" : {  "method" : "sha512", "value" : "z4PhNX7vuL3xVChQ1m2AB9Yg5AULVxXcg/SpIdNs6c5H0NE8XYXysP+DGNKHfuwvY7kxvUdBeoGlOD\nJ6+SfaPg=="  }, "source" : "A", "size" : "0", "atime" : "20240118T225653.4733203", "mtime" : "20240118T225814.87772124", "mode" : "664" }  

2024-01-18 17:58:27,057 [INFO] published: { "pubTime":"20240118225827.04942486", "baseUrl":"http://localhost:8090", "relPath":"home/peter/sarra_devdocroot/hoho/lala_wrapper.txt", "topic":"v03.post.home.peter.sarra_devdocroot.hoho", "identity":{  "method" : "sha512", "value" : "z4PhNX7vuL3xVChQ1m2AB9Yg5AULVxXcg/SpIdNs6c5H0NE8XYXysP+DGNKHfuwvY7kxvUdBeoGlOD\nJ6+SfaPg=="  } , "mtime":"20240118225814.87772124", "atime":"20240118225653.4733203", "mode":"0664", "size":"0"}
2024-01-18 17:58:27,057 [DEBUG] 620464 cpost post broker closing.
2024-01-18 17:58:27,057 [DEBUG] cpost post broker closed.
2024-01-18 17:58:27,057 [DEBUG] realpath_adjust -1, /home/peter/.cache/sr3/cpost/t_dd1_f00 -> /home/peter/.cache/sr3/cpost/t_dd1_f00 
2024-01-18 17:58:27,058 [ERROR] posting outside of post_baseDir (/home/peter/sarra_devdocroot) invalid path: /home/peter/.cache/sr3/cpost/t_dd1_f00/i01.pid
2024-01-18 17:58:27,058 [DEBUG] 620464 shim post broker closing.
2024-01-18 17:58:27,058 [DEBUG] shim post broker closed.

```

The this post is clearly using the expected configuration, as the source set to A in the published message shows.  The client observed that it looked like the SR_POST_CONFIG (shim) post setting, and the posting configuration provided on the command line, seemed to be mixed... or one was applying to the other.

So far, this is working fine. haven't reproduced a thing.

**Comment by petersilva** [comment]
Thing I observed that I don't like.  It looks like the C stuff requires the file name (complete path) as argument to find the configuration.  The python tool doesn't even accept full paths, but rather symbolic <component>/<cfg>  specifications.  It is weird that the C and python are diametrically opposed in that respect.  Opened #142

**Comment by petersilva** [comment]
just repeated the test on an rhel vm.  same... no problem at all.

**Comment by racetted** [comment]
In your sr3_cpost, we see that the shim library gets read: 

```
2024-01-18 17:58:27,044 [DEBUG] sr_config_read about to open: /home/peter/.config/sr3/post/shim_f63.conf
2024-01-18 17:58:27,044 [DEBUG] sr_shim 3.23.11p3 settings: action=foreground hostname=blacklab config_name=NONE log_level=5
```

is that expected for a process not creating files? 

In addition, my theory is that the shim config is read, and that the log level from the original post configuration gets squashed for the remainder of the process, so info message announcing the publishing of the post can be filtered if the shim configuration has a stricter log level.

**Comment by petersilva** [comment]
good point. looking.

**Comment by petersilva** [comment]
sr_util, where the log message routines are implemented, has a bunch of static variables, that include loglevel, logmode, etc... that are global.  If a single process calls the routines once for the library, and a second time parsing command line options, it might merge them.... or rather... the last one likely wins...

**Comment by petersilva** [comment]
OK it still wasn't fixed at first, after another patch to specifically address loglevel, it now seems to pass the test above (shim stuff loglevel is no longer debug.)

the latest patch also changes output slightly:

```
2024-02-26 15:25:56,270 [INFO] shim published: { "pubTime":"20240226202556.26393323", "baseUrl":"ftp://anonymous@localhost:2121/", "relPath":"hoho/.libsr3shim.c.swx", "topic":"v03.post.hoho", "fileOp" : { "remove":""}}

2024-02-26 15:25:56,270 [INFO] cpost published: { "pubTime":"20240226202556.26393323", "baseUrl":"ftp://anonymous@localhost:2121/", "relPath":"hoho/.libsr3shim.c.swx", "topic":"v03.post.hoho", "fileOp" : { "remove":""}}

```

prefixing "published:" by the program doing the call, either shim library or cpost.

**Comment by racetted** [comment]
So to confirm, the log level will be consistent with the configuration set for each tool (cpost and shim) without clashing globals. 

Did you find out why there is a shim post in the first place for a cpost call?

**Comment by petersilva** [comment]
hold on... It's failing some other tests I had not run.

To address your question:
```
fractal% /bin/echo hoho >>lala
2024-02-26 15:25:08,213 [INFO] amqp_consumer option not implemented, ignored.
2024-02-26 15:25:08,213 [INFO] logEvents option not implemented, ignored.
2024-02-26 15:25:08,224 [INFO] shim published: { "pubTime":"20240226202508.21655463", "baseUrl":"ftp://anonymous@localhost:2121/", "relPath":"hoho/lala", "topic":"v03.post.hoho", "identity":{  "method" : "sha512", "value" : "/l01soI1pdoBSskpXyIHvElRRKXCXnZGOwmgW99iuiQzmFIzA/rBdo+0DXWeN4wlxQ5dCvarK7GwtI\ngdCdoeww=="  } , "mtime":"20240226202508.21050133", "atime":"20240226200650.34609877", "mode":"0664", "size":"30"}
fractal% sr3_cpost -c ~/.config/sr3/post/t_dd1_f00.conf -p lala*
2024-02-26 15:25:14,169 [INFO] amqp_consumer option not implemented, ignored.
2024-02-26 15:25:14,169 [INFO] logEvents option not implemented, ignored.
2024-02-26 15:25:14,169 [ERROR] Environment variable not set: SAMPLEDATA
2024-02-26 15:25:14,169 [INFO] set option not implemented, ignored.
2024-02-26 15:25:14,169 [INFO] set option not implemented, ignored.
2024-02-26 15:25:14,170 [DEBUG] cpost 3.24.01 settings: action=foreground hostname=fractal config_name=t_dd1_f00 log_level=6
2024-02-26 15:25:14,170 [DEBUG] 	follow_symlinks=no realpath: Adjust=0 Filter=yes Post=no
2024-02-26 15:25:14,170 [DEBUG] 	force_polling=off sleep=0 expire=180 housekeeping=300 sanity_log_dead=450 nodupe_ttl=0 nodupe_fileAgeMin=0 nodupe_fileAgeMax=2.592e+06
2024-02-26 15:25:14,170 [DEBUG] 	cache_file=off cache_basis=path, accept_unmatch=on messageRateMax=0
2024-02-26 15:25:14,170 [DEBUG] 	fileEvents=006f directory=(null) queuename=(null) logReject=on sum=s
2024-02-26 15:25:14,170 [DEBUG] 	statehost=no v2compatRenameDoublePost=no
2024-02-26 15:25:14,170 [DEBUG] 	message_ttl=0 post_exchange=xsarra post_exchange_split=0 post_exchangeSuffix=(null)
2024-02-26 15:25:14,170 [DEBUG] 	source=A to=localhost post_baseUrl=http://localhost:8090/ post_baseDir=(null)
2024-02-26 15:25:14,170 [DEBUG] 	topicPrefix=v03 post_topicPrefix=v03.post, pid=-1
2024-02-26 15:25:14,170 [DEBUG] man sr3_cpost(1) for more information
2024-02-26 15:25:14,172 [DEBUG] 3.24.01 connected to post broker amqp://tfeed@localhost:5672/#xsarra
2024-02-26 15:25:14,172 [DEBUG] declaring exchange amqp://tfeed@localhost:5672/#xsarra
2024-02-26 15:25:14,172 [DEBUG] sr_config_read about to open: /home/peter/.config/sr3/default.conf
2024-02-26 15:25:14,172 [INFO] amqp_consumer option not implemented, ignored.
2024-02-26 15:25:14,172 [INFO] logEvents option not implemented, ignored.
2024-02-26 15:25:14,172 [DEBUG] sr_config_read about to open: /home/peter/.config/sr3/post/shim_f63.conf
2024-02-26 15:25:14,174 [ERROR] shim posting outside of post_baseDir (/home/peter/sarra_devdocroot) invalid path: /home/peter/.cache/sr3/cpost/t_dd1_f00/i01.pid
2024-02-26 15:25:14,174 [INFO] cpost 3.24.01 config: t_dd1_f00, pid: 216851, starting
2024-02-26 15:25:14,182 [INFO] cpost published: { "pubTime":"20240226202514.17449127", "baseUrl":"http://localhost:8090/", "relPath":"home/peter/sarra_devdocroot/hoho/lala", "topic":"v03.post.home.peter.sarra_devdocroot.hoho", "identity":{  "method" : "sha512", "value" : "/l01soI1pdoBSskpXyIHvElRRKXCXnZGOwmgW99iuiQzmFIzA/rBdo+0DXWeN4wlxQ5dCvarK7GwtI\ngdCdoeww=="  } , "mtime":"20240226202508.21050133", "atime":"20240226202508.21450138", "mode":"0664", "size":"30"}
2024-02-26 15:25:14,183 [ERROR] shim posting outside of post_baseDir (/home/peter/sarra_devdocroot) invalid path: /home/peter/.cache/sr3/cpost/t_dd1_f00/i01.pid
fractal% vi libsr3shim.c
2024-02-26 15:25:56,270 [INFO] shim published: { "pubTime":"20240226202556.26393323", "baseUrl":"ftp://anonymous@localhost:2121/", "relPath":"hoho/.libsr3shim.c.swx", "topic":"v03.post.hoho", "fileOp" : { "remove":""}}
2024-02-26 15:25:56,275 [INFO] shim published: { "pubTime":"20240226202556.27015464", "baseUrl":"ftp://anonymous@localhost:2121/", "relPath":"hoho/.libsr3shim.c.swp", "topic":"v03.post.hoho", "fileOp" : { "remove":""}}

```

This is what it is doing now... but some other tests fail. I have to look more...

**Comment by petersilva** [comment]
The test harness needed updating... the code itself is fine... patch inbound...
